docs/adrs/0002-uds-core-functional-layers.md

```diff
@@ -0,0 +1,73 @@
+# ADR: Layered Architecture for UDS Core within a Monorepo
+
+Date: 2024-07-17
+
+## Status
+
+Accepted
+
+## Context
+
+Historically, UDS Core intentionally only published a single official Zarf Package to ensure a standard baseline across all UDS environments. We learned form the complex buffet menu-style of Big Bang that it is better to form opinions and test them than to allow unlimited configuration possibilities. We have also found the monorepo structure of UDS Core to be be much simpler for developing UDS Core. However, through user feedback, we have seen cases where more optionality is warranted, but having to modify UDS CLI to accommodate disabling portions of UDS Core is not ideal. We also have a need for an edge-compatible version of UDS Core that is more lightweight and does not include all the components of the full UDS Core package.
+
+Today, we publish both the official UDS Core package and a UDS Slim Dev package, in addition to dev and demo UDS Bundles. The UDS Slim Dev package is a stripped-down version of UDS Core that includes limited services meant for testing UDS Packages. We have also learned that some teams outside of UDS need similar functionality, but slightly different parts of UDS Core, such as keycloak + authsvc.
+
+## Decision
+
+We will replace the existing 2-package (one dev and one prod) structure with a layered set of packages within the existing monorepo. This will allow users to choose the components they need for their specific use case, while maintaining a single source of truth for all UDS Core components. We will not break these out by applications as Big Bang did, but instead by layers of functionality for related/integrated components. This allows more flexibility for which functionality is provided by UDS Core and which is provided by other external packages.
+
+The functional layers will be:
+
+1. UDS Core Base (Istio + Pepr w/UDS Operator & UDS Policies)
+2. UDS Core Identity and Authorization (Keycloak + AuthService)
+3. UDS Core Metrics (Metrics Server)
+4. UDS Core Monitoring (Prometheus + Grafana)
+5. UDS Core Logging (Promtail + Loki)
+6. UDS Core Runtime Security (NeuVector)
+7. UDS Core Backup and Restore (Velero)
+8. UDS Core (combination of all layers)
+
+These layers can be combined as needed, with UDS Core Base serving as the foundation for all deployments. The UDS Core package will continue to offer the complete suite of tools. The only change in the UDS Core repo will be the addition of new package definitions for each functional layer. Versioning will be managed at the monorepo level, with each layer being versioned in lockstep with the rest of UDS Core.
+
+We will still publish the UDS Slim Dev bundle for testing UDS Packages. This will be a combination of UDS Core Base, UDS Core Identity and Authorization, and UDS-K3d. We will not publish other bundles beyond UDS Slim Dev and UDS Core Demo. Only the UDS Core zarf package will publish combined layers.
```

Comment:
For this will there still be the CRs for Prometheus things? (to allow the monitor key to work / create resources from the operator)

Reply:
tbd I think, want to do whatever makes sense here, realize rn we have to push this at the beginning of UDS Core, so maybe it stays in the base layer, but needs more dialog

Reply:
For my 2 pennies I'd advocate having that in there (or having an "operator dev" layer) that allows the full functionality of the UDS operator without needing everything deployed (which could push up against resources)

Reply:
the operator should be smart enough to handle this imo

Reply:
Yeah I think in most cases, including `monitor` keys/pieces, we should just have the operator logic be more intelligent about looking up current cluster state and skipping over pieces that aren't applicable. Imo there is little benefit in having the CRs created if nothing is "using them" (i.e. the CRD exists but the prometheus-operator isn't deployed). I recognize the desire to CI test bits of functionality like monitoring but if you only have the CRDs deployed you're only testing one aspect (ex: servicemonitor creation) not the full picture (ex: servicemonitor actually works).

Reply:
Yeah I could get behind that too - we mostly assume this will #work today but it definitely makes more sense to validate in fuller contexts